### PR TITLE
Feature(brisque): Added BRISQUE score

### LINKED ISSUE #69: Metric: BRISQUE
**Is your feature request related to a problem? Please describe.**
Implement [BRASQUE](https://live.ece.utexas.edu/publications/2012/TIP%20BRISQUE.pdf) metric or adapt [existing implementation](https://github.com/bukalapak/pybrisque)

### PR DESCRIPTION
**Pull request by denproc**
Closes #69 

## Proposed Changes
- Feature(brisque): Added BRISQUE score
- Tests(brisque): Added tests for BRISQUE
- Docs(brisque): Updated README.md
- Minor(utils): Added parameter to enable 5d check at validation of input

### CODE REVIEW

**Comment by denproc** [comment]
Ready for review, @zakajd @snk4tr

File: tests/test_brisque.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                                                                             f'got {score} and {score_baseline}'
+
+
+def test_brisque_values_3d(prediction_RGB: torch.Tensor) -> None:
+    score = brisque(prediction_RGB, reduction='none')
+    score_baseline = torch.tensor([BRISQUE().get_score(img.squeeze().permute(1, 2, 0).numpy()[..., ::-1])
+                                   for img in prediction_RGB])
+    assert torch.isclose(score, score_baseline, rtol=2e-4).all(), f'Expected values to be equal to ' \
+                                                                  f'baseline prediction.' \
+                                                                  f'got {score} and {score_baseline}'
```
Comment (snk4tr) [inline tests/test_brisque.py:77]: Naming: same as above

File: tests/test_brisque.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        BRISQUELoss()(prediction_grey)
+    except Exception as e:
+        pytest.fail(f"Unexpected error occurred: {e}")
+
+
+def test_brisque_loss_if_works_with_3d(prediction_RGB: torch.Tensor) -> None:
+    try:
+        BRISQUELoss()(prediction_RGB)
+    except Exception as e:
+        pytest.fail(f"Unexpected error occurred: {e}")
```
Comment (snk4tr) [inline tests/test_brisque.py:92]: Naming: same as above

File: photosynthesis_metrics/brisque.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+            .. [1] Anish Mittal et al. "No-Reference Image Quality Assessment in the Spatial Domain",
+            https://live.ece.utexas.edu/publications/2012/TIP%20BRISQUE.pdf
+        """
+    _validate_input(input_tensors=x)
+    x = _adjust_dimensions(input_tensors=x)
+    if data_range == 255:
+        x = x / 255
+
+    if x.size(1) == 3:
+        x = torch.sum(x * torch.tensor([0.299, 0.587, 0.114]).view(1, -1, 1, 1), dim=1, keepdim=True)
```
[the remark this answers is not in the document]
Comment (denproc) [reply photosynthesis_metrics/brisque.py:162]: The values are taken from RGB to Grey image transformation. These values are constant, no need in parametrisation.

File: photosynthesis_metrics/brisque.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+    rho = sigma_sq / E ** 2
+
+    indexes = (rho - r_table).abs().argmin(dim=-1)
+    solution = gamma[indexes]
+
+    return solution, sigma
+
+
+def _aggd(x: torch.Tensor) -> tuple:
+    # It works
```
[the remark this answers is not in the document]
Comment (denproc) [reply photosynthesis_metrics/brisque.py:36]: This is the common name for (asymmetric) generalized Gaussian distribution. The function estimate its parameters. I'll rename it to `_ggd_parameters` and  `_aggd_parameters` .

File: photosynthesis_metrics/brisque.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+                                   [0.000975, 0.470984], [0.249, 1.555], [-0.135687, 0.100858],
+                                   [0.000174, 0.684173], [0.000913, 0.534174], [0.258, 1.561],
+                                   [-0.143408, 0.100486], [0.000179, 0.685696], [0.000888, 0.536508],
+                                   [0.471, 3.264], [0.012809, 0.703171], [0.218, 1.046],
+                                   [-0.094876, 0.187459], [1.5e-005, 0.442057], [0.001272, 0.40803],
+                                   [0.222, 1.042], [-0.115772, 0.162604], [1.6e-005, 0.444362],
+                                   [0.001374, 0.40243], [0.227, 0.996],
+                                   [-0.117188, 0.09832299999999999], [3e-005, 0.531903],
+                                   [0.001122, 0.369589], [0.228, 0.99], [-0.12243, 0.098658],
+                                   [2.8e-005, 0.530092], [0.001118, 0.370399]])
```
[the remark this answers is not in the document]
Comment (denproc) [reply photosynthesis_metrics/brisque.py:113]: Feature range is taken from [official implementation of BRISQUE on MATLAB](https://live.ece.utexas.edu/research/Quality/index_algorithms.htm).

File: photosynthesis_metrics/brisque.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+    dist = (features - sv).pow(2).sum(dim=1)
+    return torch.exp(- dist * gamma)
+
+
+def _score_svr(features: torch.Tensor) -> torch.Tensor:
+
+    url = 'https://drive.google.com/uc?export=download&id=1e1F4jLh2-P3mqZ20GYr9ItgLma6999CX'
+    sv_coef, sv = load_url(url, map_location=features.device)
+    gamma = 0.05
+    rho = -153.591
```
[the remark this answers is not in the document]
Comment (denproc) [reply photosynthesis_metrics/brisque.py:133]: These values are taken from SVM model parameters from official implementation of [BRISQUE on MATLAB](https://live.ece.utexas.edu/research/Quality/index_algorithms.htm).

File: photosynthesis_metrics/brisque.py [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+    if data_range == 255:
+        x = x / 255
+
+    if x.size(1) == 3:
+        x = torch.sum(x * torch.tensor([0.299, 0.587, 0.114]).view(1, -1, 1, 1), dim=1, keepdim=True)
+    features = []
+    num_of_scales = 2
+    for iteration in range(num_of_scales):
+        features.append(_nss(x, kernel_size, kernel_sigma))
+        x = F.interpolate(x, scale_factor=0.5, mode='nearest')
```
[the remark this answers is not in the document]
Comment (denproc) [reply photosynthesis_metrics/brisque.py:167]: `scale_parameter` is specified in the paper. Also, it affect the number of features to be used as an input of SVM model. The paper and official implementation describe only 36 features derived from 2 scales. No need in parametrisation.

[the remark this answers is not in the document]
Comment (denproc) [reply photosynthesis_metrics/brisque.py:167]: In case of the `mode` it seems reasonable.

File: photosynthesis_metrics/utils.py [hunk 8]
```diff
@@ -25,18 +25,22 @@ def _adjust_dimensions(input_tensors: Union[torch.Tensor, Tuple[torch.Tensor, to
...
     assert 0 < len(input_tensors) < 3, f'Expected one or two input tensors, got {len(input_tensors)}'
 
+    max_size = 5 if enable_5d else 4
+
     for tensor in input_tensors:
         assert isinstance(tensor, torch.Tensor), f'Expected input to be torch.Tensor, got {type(tensor)}.'
-        assert 1 < tensor.dim() < 6, f'Input images must be 2D, 3D or 4D tensors, got images of shape {tensor.size()}.'
-        if tensor.dim() == 5:
+        assert 1 < tensor.dim() <= max_size, f'Input images must be 2D, 3D or 4D tensors, ' \
+                                             f'got images of shape {tensor.size()}.'
```
[the remark this answers is not in the document]
Comment (denproc) [reply photosynthesis_metrics/utils.py:42]: I'll wait for #78

File: photosynthesis_metrics/utils.py [hunk 9]
```diff
@@ -25,18 +25,22 @@ def _adjust_dimensions(input_tensors: Union[torch.Tensor, Tuple[torch.Tensor, to
...
+        if tensor.dim() == 5 and enable_5d:
```
[the remark this answers is not in the document]
Comment (denproc) [reply photosynthesis_metrics/utils.py:43]: I'll wait for #78

File: tests/test_brisque.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+@pytest.fixture(scope='module')
+def prediction_RGB() -> torch.Tensor:
+    return torch.rand(3, 3, 256, 256)
+
+
+# ================== Test function: `brisque` ==================
+def test_brisque_raises_if_wrong_type_input() -> None:
+    with pytest.raises(AssertionError):
+        brisque(None)
```
[the remark this answers is not in the document]
Comment (denproc) [reply tests/test_brisque.py:21]: #78 introduces these checks, I suggest to remove it from here.

File: tests/test_brisque.py [hunk 11]
```diff
@@ -0,0 +1,103 @@
+import torch
+import pytest
+from libsvm import svmutil  # noqa: F401
```
[the remark this answers is not in the document]
Comment (denproc) [reply tests/test_brisque.py:3]: - `pybrisque` (`brisque`) [library](https://github.com/bukalapak/pybrisque/blob/master/brisque/brisque.py) uses `import svmutil`.
- `svmutil` is installed as a part of `libsvm` from pypi. As a result `brisque` cannot import  `svmutil`directly.
- The resolve the problem I proposed to import `svmutil` manually.

Similar problem is also described [here](https://github.com/ocampor/notebooks/issues/10)

File: tests/test_brisque.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (snk4tr) [reply tests/test_brisque.py:21]: Please do it

**Comment by denproc** [comment]
ready for second review. Also added new tag into setup.

**Comment by snk4tr** [comment]
Will have a look tomorrow

**Comment by denproc** [comment]
Additional restriction was introduced to requirements.txt. `argmin` and `argmax` are corrupted and cannot be used for back propagation in pytorch v.1.5.0.

More info: https://github.com/pytorch/pytorch/pull/39257
